### PR TITLE
Settings bounds

### PR DESCRIPTION
**Pull request by danieljaeim**
Added functionality to enable users to have controls over their math bounds in screenshot captures (#7 )

I started by creating additional inputs for top, bottom, left, and right bounds that correspond to the edges of the image returned from the[ asyncScreenshot()](https://www.desmos.com/api/v1.2/docs/index.html#GraphingCalculator.asyncScreenshot) method.
I also added a dropdown to denote the user's choice of image setting between ('contain', 'stretch, 'preserveY', preserveX').

59388615-1955c780-8d21-11e9-89ab-fe6019ac6a71

I then added the left, right, top, and bottom values from our inputs to our redux-store, passing them onto the Settings.js Component through the SettingsContainer. Now our settingsComponent could update and refer to our bound values from the store.

I then updated the settings action creator to account for the UPDATE_BOUNDS_SETTING, which adjusts our store's left, right, top, and bottom values as the user changes them live.

Now when we fetch an image from the API through the requestFrame() method, we also pass in an object 'mathBounds', that act as additional parameters for the async ScreenShot() method, accounting for errors such as if our left bound is greater than our right bound, or whether our bottom bound is greater than our top bound.

I finally adjusted the errors in requestFrame, to account for invalidBounds. Now when a user tries to capture the graph and send a request to the API, it validates whether the bounds follow protocol
(left bound is less than right, bottom bound is less than top), and flashes an error to the top-banner if an error occurs.

I also wrote tests for these as well, and they work in making sure our bounds are correct :)

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -5,6 +5,7 @@
 /.pnp
 .pnp.js
 
+
```
Comment (colebillys19) [inline .gitignore:8]: Remove this extra line

File: src/components/App.test.js [hunk 2]
```diff
@@ -3,6 +3,6 @@ import App from './App';
 
 describe('App', () => {
   it('renders without crashing', () => {
-    renderWithRedux(<App />);
+    let wrapper = global.renderWithRedux(<App />);
```
Comment (colebillys19) [inline src/components/App.test.js:6]: Probably best to scrap any testing related code from this PR. Testing stuff can live in a separate one.

File: src/components/Sidebar.js [hunk 3]
```diff
@@ -53,12 +70,14 @@ class Sidebar extends Component {
         <SidebarButton icon="camera" onClick={this.handleRequestFrame} />
 
         <SidebarButton
+          data-testid="burst"
```
Comment (colebillys19) [inline src/components/Sidebar.js:73]: Remove this for now

File: src/components/Sidebar.js [hunk 4]
```diff
@@ -53,12 +70,14 @@ class Sidebar extends Component {
...
           icon="burst"
           expanded={expandedPane === panes.BURST}
           onClick={this.handleToggleBurst}
         />
 
         <SidebarButtonWithBadge
+          data-testid="preview"
```
Comment (colebillys19) [inline src/components/Sidebar.js:80]: Remove this for now

File: src/components/Sidebar.js [hunk 5]
```diff
@@ -68,6 +87,7 @@ class Sidebar extends Component {
         />
 
         <SidebarButton
+          data-testid="settings"
```
Comment (colebillys19) [inline src/components/Sidebar.js:90]: Remove this for now

File: src/components/Sidebar.js [hunk 6]
```diff
@@ -77,6 +97,7 @@ class Sidebar extends Component {
 
         {!!gifData.length && (
           <SidebarButtonWithBadge
+            data-testid="download"
```
Comment (colebillys19) [inline src/components/Sidebar.js:100]: Remove this for now

File: src/components/__tests__/sidebar.test.js [hunk 7]
```diff
@@ -0,0 +1,19 @@
+import React from 'react';
```
Comment (colebillys19) [inline src/components/__tests__/sidebar.test.js:1]: Get rid of this file and folder for now

File: src/components/Settings.js [hunk 8]
```diff
@@ -1,25 +1,45 @@
...
     super(props);
     this.handleInputUpdate = this.handleInputUpdate.bind(this);
+    this.handleStrategyUpdate = this.handleStrategyUpdate.bind(this);
   }
 
   handleInputUpdate(evt) {
     const {
       target: { name, value, checked }
     } = evt;
+
```
Comment (colebillys19) [inline src/components/Settings.js:17]: Extra whitespace

File: src/components/Settings.js [hunk 9]
```diff
@@ -38,6 +58,7 @@ class Settings extends Component {
           value={isNaN(width) ? '' : width}
           onChange={this.handleInputUpdate}
         />
+
```
Comment (colebillys19) [inline src/components/Settings.js:61]: Extra whitespace

File: src/components/Settings.js [hunk 10]
```diff
@@ -49,6 +70,7 @@ class Settings extends Component {
           value={isNaN(height) ? '' : height}
           onChange={this.handleInputUpdate}
         />
+
```
Comment (colebillys19) [inline src/components/Settings.js:73]: Extra whitespace

File: src/components/Settings.js [hunk 11]
```diff
@@ -60,6 +82,7 @@ class Settings extends Component {
           value={isNaN(interval) ? '' : interval}
           onChange={this.handleInputUpdate}
         />
+
```
Comment (colebillys19) [inline src/components/Settings.js:85]: Extra whitespace

File: src/lib/calc-helpers.js [hunk 12]
```diff
@@ -37,5 +37,6 @@ export const setSliderByIndex = (idx, val) => {
   if (!match) return notASlider(idx);
 
   const identifier = match[1];
+
```
Comment (colebillys19) [inline src/lib/calc-helpers.js:40]: Extra whitespace

File: src/components/App.js [hunk 13]
```diff
@@ -30,6 +30,8 @@ class App extends Component {
       CALCULATOR_OPTIONS
     );
 
+    //allows us to see calculator object in window
```
Comment (japamat0) [inline src/components/App.js:33]: I imagine this was put in for development purposes, should be removed for PR

File: src/components/Settings.css [hunk 14]
```diff
@@ -38,3 +38,12 @@
...
 }
+
+.Settings-dropdown {
+  height: 33px;
+  width: 114px;
+  margin-bottom: 15px;
+  margin-top: 5px;
+  border: 1px solid #000;
+  font-size: 1em;
+}
```
Comment (japamat0) [inline src/components/Settings.css:49]: add `outline: none !important;` to prevent dropdown outline on focus

File: src/lib/error-messages.js [hunk 15]
```diff
@@ -64,3 +64,43 @@ export const badSettingsInput = errors => {
...
+      case 'bottom':
+        errorString.push('Bottom Bound must be a valid integer.');
+        break;
+      case 'left':
+        errorString.push('Left Bound must be a valid integer.');
+        break;
+      case 'right':
+        errorString.push('Right Bound must be a valid integer.');
+        break;
+    }
```
Comment (japamat0) [inline src/lib/error-messages.js:102]: this `switch` function needs a `default` case

File: src/lib/input-helpers.js [hunk 16]
```diff
@@ -35,3 +39,17 @@ export const getSettingsErrors = inputs => {
...
+  const errors = {};
+  const { left, right, top, bottom } = inputs;
+
+  if (left >= right) errors['leftless'] = true;
+  if (bottom >= top) errors['bottomless'] = true;
+  if (isNaN(top)) errors['top'] = true;
+  if (isNaN(bottom)) errors['bottom'] = true;
+  if (isNaN(left)) errors['left'] = true;
+  if (isNaN(right)) errors['right'] = true;
+
```
Comment (japamat0) [inline src/lib/input-helpers.js:53]: 49-52 can be a `for... in` loop:
```
for (let cur in inputs) {
  if (isNaN(inputs[cur])) errors[cur] = true;
} 
```

**Pull request merged**